volumes/miovision/sql/function-find_gaps.sql

```diff
@@ -1,116 +1,199 @@
+    FROM daily_intersections AS i
+    --add artificial data points at start and end of day to find gaps overlapping start/end.
+    CROSS JOIN (
+        VALUES
+            --catch gaps overlapping days
+            (start_date::timestamp - interval '15 minutes'),
+            (start_date::timestamp + interval '1 day')
+    ) AS bins(datetime_bin)
+
+    --group by in next step takes care of duplicates
+    UNION ALL
+
+    --the distinct datetime bins which did appear in the day's data. 
+    SELECT DISTINCT
+        intersection_uid,
+        datetime_bin,
+        --need to reduce gap length by 1 minute for real data since that minute contains data
+        interval '1 minute' AS gap_adjustment
+    FROM miovision_api.volumes
+    WHERE
+        datetime_bin >= start_date - interval '15 minutes'
+        AND datetime_bin < start_date + interval '1 day'
+),
+
+--looks at sequential bins to identify breaks larger than 1 minute.
+bin_times AS (
+    SELECT 
+        intersection_uid,
+		--sum works because between 0 and 1, we want 1 (implies real data)
+        datetime_bin + SUM(gap_adjustment) AS gap_start,
+        date_part('hour', datetime_bin) AS time_bin,
+        LEAD(datetime_bin, 1) OVER (PARTITION BY intersection_uid ORDER BY datetime_bin) AS gap_end,
+        LEAD(datetime_bin, 1) OVER (PARTITION BY intersection_uid ORDER BY datetime_bin)
+            - datetime_bin > interval '1 minute' AS bin_break, --True means gap between bins is larger than 1 minute
+        --weekend is needed to determine which gapsize_lookup to use. 
+        CASE
+            WHEN date_part('isodow', datetime_bin) <= 5
+                AND hol.holiday IS NULL THEN False
+            ELSE True
+        END as weekend
+    FROM fluffed_data
+    LEFT JOIN ref.holiday AS hol ON hol.dt = fluffed_data.datetime_bin::date
+    GROUP BY
+        intersection_uid, 
+        datetime_bin,
+        weekend
+),
+
+gaps AS (
+	--find gaps longer than the threshold from gapsize_lookup and cross join with
+    --generate_series output to enable easily joining with 15 minute datetime_bins.
+	INSERT INTO miovision_api.unacceptable_gaps(
+		intersection_uid, gap_start, gap_end, gap_minutes_total, allowable_total_gap_threshold, 
+        datetime_bin, gap_minutes_15min, avg_historical_total_vol, avg_historical_veh_vol
+	)    
+    SELECT
+		bt.intersection_uid,
+		bt.gap_start,
+		bt.gap_end,
+		gm.gap_minutes_total,
+        gl.gap_tolerance AS allowable_total_gap_threshold,
+        bins.datetime_bin,		
+        round(gm.gap_minutes_15min) AS gap_minutes_15min,
+        --avg historical volumes during the gaps to use for imputation
+        --or decision on summary data quality
+        round(gl.avg_vol * gm.gap_minutes_15min / 60) AS avg_historical_total_vol,
+        round(gl.avg_vehicle_vol * gm.gap_minutes_15min / 60) AS avg_historical_veh_vol
+	FROM bin_times AS bt
+    --match gaps to the 15 minute bins they intersect
+    LEFT JOIN generate_series(
```

Comment:
I think these `LEFT JOIN`s end up effectively being an `INNER JOIN` given the `WHERE` clause depends on `gm` which depends on the two previous joins. Would be more explicit to make them `INNER`